zebrad/tests/common/test_type.rs

```diff
@@ -0,0 +1,316 @@
+use std::{env, path::PathBuf, time::Duration};
+
+use zebra_test::{command::NO_MATCHES_REGEX_ITER, prelude::*};
+use zebrad::config::ZebradConfig;
+
+use super::{
+    cached_state::ZEBRA_CACHED_STATE_DIR,
+    config::{default_test_config, random_known_rpc_port_config},
+    failure_messages::{
+        LIGHTWALLETD_EMPTY_ZEBRA_STATE_IGNORE_MESSAGES, LIGHTWALLETD_FAILURE_MESSAGES,
+        PROCESS_FAILURE_MESSAGES, ZEBRA_FAILURE_MESSAGES,
+    },
+    launch::{LIGHTWALLETD_DELAY, LIGHTWALLETD_FULL_SYNC_TIP_DELAY, LIGHTWALLETD_UPDATE_TIP_DELAY},
+    lightwalletd::LIGHTWALLETD_DATA_DIR,
+    sync::FINISH_PARTIAL_SYNC_TIMEOUT,
+};
+
+use TestType::*;
+
+/// The type of integration test that we're running.
+#[derive(Copy, Clone, Debug, Eq, PartialEq)]
+pub enum TestType {
+    /// Launch with an empty Zebra and lightwalletd state.
+    LaunchWithEmptyState,
+
+    /// Do a full sync from an empty lightwalletd state.
+    ///
+    /// This test requires a cached Zebra state.
+    //
+    // Only used with `--features=lightwalletd-grpc-tests`.
+    #[allow(dead_code)]
+    FullSyncFromGenesis {
+        /// Should the test allow a cached lightwalletd state?
+        ///
+        /// If `false`, the test fails if the lightwalletd state is populated.
+        allow_lightwalletd_cached_state: bool,
+    },
+
+    /// Sync to tip from a lightwalletd cached state.
+    ///
+    /// This test requires a cached Zebra and lightwalletd state.
+    UpdateCachedState,
+
+    /// Launch `zebrad` and sync it to the tip, but don't launch `lightwalletd`.
+    ///
+    /// If this test fails, the failure is in `zebrad` without RPCs or `lightwalletd`.
+    /// If it succeeds, but the RPC tests fail, the problem is caused by RPCs or `lightwalletd`.
+    ///
+    /// This test requires a cached Zebra state.
+    UpdateZebraCachedStateNoRpc,
+
+    /// Launch `zebrad` and sync it to the tip, but don't launch `lightwalletd`.
+    ///
+    /// This test requires a cached Zebra state.
+    UpdateZebraCachedState,
```

Comment:
Nitpick: The variant name seems a bit ambiguous, how about:

```suggestion
    /// Launch `zebrad`, open a RPC port, and sync it to the tip, but don't launch `lightwalletd`.
    ///
    /// This test requires a cached Zebra state.
    #[allow(dead_code)]
    UpdateZebraCachedStateWithRpc,
```

Reply:
Renamed in [`d72ea40`](https://github.com/ZcashFoundation/zebra/pull/5589/commits/d72ea40650530d1dfa5e3bdd73d9d146eb28fba8)